src/AutoHashEqualsCached.jl

```diff
@@ -0,0 +1,220 @@
+  implementation of `T._show_default` for types `T` annotated with
+  `@auto_hash_equals_cached`.  This is ultimately used in the implementation of
+  `Base.show`.  This specialization ensures that showing circular data structures does not
+  result in infinite recursion.
+"""
+function _show_default_auto_hash_equals_cached(io::IO, @nospecialize(x))
+    t = typeof(x)
+    show(io, Base.inferencebarrier(t)::DataType)
+    print(io, '(')
+    recur_io = IOContext(io, Pair{Symbol,Any}(:SHOWN_SET, x),
+                         Pair{Symbol,Any}(:typeinfo, Any))
+    if !Base.show_circular(io, x)
+        for i in 1:nfields(x)
+            f = fieldname(t, i)
+            if (f == :_cached_hash)
+                continue
+            elseif i > 1
+                print(io, ", ")
+            end
+            if !isdefined(x, f)
+                print(io, Base.undef_ref_str)
+            else
+                show(recur_io, getfield(x, i))
+            end
+        end
+    end
+    print(io,')')
+end
+
+"""
+An exception used for reporting errors.
+"""
+struct AutoHashEqualsException <: Exception
+    msg::String
+end
+error(msg::String) = throw(AutoHashEqualsException(msg))
+
+"""
+Find the first struct declaration buried in the Expr.
+"""
+get_struct_decl(typ) = nothing
+function get_struct_decl(typ::Expr)
+    if typ.head == :struct
+        return typ
+    elseif typ.head == :macrocall
+        return get_struct_decl(typ.args[3])
+    elseif typ.head == :block
+        # get the first struct decl in the block
+        for x in typ.args
+            if x.head == :macrocall || x.head == :struct || x.head == :block
+                result = get_struct_decl(x)
+                if (result !== nothing)
+                    return result
+                end
+            end
+        end
+    end
+
+    error("macro @auto_hash_equals_cached should only be applied to a struct")
+end
+
+unpack_name(node) = node
+function unpack_name(node::Expr)
+    if node.head == :macrocall
+        unpack_name(node.args[3])
+    elseif node.head in (:(<:), :(::))
+        unpack_name(node.args[1])
+    else
+        node
+    end
+end
+
+unpack_type_name(n::Symbol) = (n, n, nothing)
+function unpack_type_name(n::Expr)
+    if n.head == :curly
+        type_name = n.args[1]
+        type_name isa Symbol ||
+            error("macro @auto_hash_equals_cached applied to type with invalid signature")
+        where_list = n.args[2:length(n.args)]
+        type_params = map(unpack_name, where_list)
+        full_type_name = Expr(:curly, type_name, type_params...)
+        (type_name, full_type_name, where_list)
+    elseif n.head == :(<:)
+        unpack_type_name(n.args[1])
+    else
+        error("macro @auto_hash_equals_cached applied to type with unexpected signature")
+    end
+end
+
+function get_fields(struct_decl::Expr)
+    member_names = Vector{Symbol}()
+    member_decls = Vector()
+
+    add_field(b) = nothing
```

Comment:
Idiomatically, methods that mutate values should be named with a `!` (as in `add_field!`). This one's local, so not a big deal.

Reply:
The reason I didn't name it that way is that it doesn't mutate its arguments.